plugins/tiddlywiki/sqlite3store/sql-functions.js

```diff
@@ -0,0 +1,217 @@
+		});
+		if(statementTiddlerExists.step()) {
+			statementTiddlerExists.reset();
+			return true;
+		} else {
+			statementTiddlerExists.reset();
+			return false;
+		}
+	};
+	var statementGetTiddler = self.db.prepare("select title, shadow, meta, text from tiddlers where title = $title order by shadow");
+	this.sqlGetTiddler = function(title) {
+		statementGetTiddler.bind({
+			$title: title
+		});
+		if(statementGetTiddler.step()) {
+			var row = statementGetTiddler.get({});
+			statementGetTiddler.reset();
+			return Object.assign({},JSON.parse(row.meta),{title: row.title, text: row.text});
+		} else {
+			statementGetTiddler.reset();
+			return undefined;
+		}
+	};
+	var statementGetShadowSource = self.db.prepare("select title, shadowsource from tiddlers where title = $title and shadow = 1");
+	this.sqlGetShadowSource = function(title) {
+		statementGetShadowSource.bind({
+			$title: title
+		});
+		if(statementGetShadowSource.step()) {
+			var row = statementGetShadowSource.get({});
+			statementGetShadowSource.reset();
+			return row.shadowsource;
+		} else {
+			statementGetShadowSource.reset();
+			return undefined;
+		}
+	};
+	var statementAllTitles = self.db.prepare("select title from tiddlers where shadow = 0 order by title");
+	this.sqlAllTitles = function() {
+		let resultRows = [];
+		while(statementAllTitles.step()) {
+			var row = statementAllTitles.get({});
+			resultRows.push(row.title);
+		}
+		statementAllTitles.reset();
+		return resultRows;
+	};
+	var statementAllShadowTitles = self.db.prepare("select title from tiddlers where shadow = 1 order by title");
+	this.sqlAllShadowTitles = function() {
+		let resultRows = [];
+		while(statementAllShadowTitles.step()) {
+			var row = statementAllShadowTitles.get({});
+			resultRows.push(row.title);
+		}
+		statementAllShadowTitles.reset();
+		return resultRows;
+	};
+	var statementEachTiddler = self.db.prepare("select title, meta, text from tiddlers where shadow = 0 order by title");
+	this.sqlEachTiddler = function(callback) {
+		while(statementEachTiddler.step()) {
+			var row = statementEachTiddler.get({}),
+				tiddlerFields = Object.assign({},JSON.parse(row.meta),{title: row.title, text: row.text});
+			callback(tiddlerFields,row.title);
+		}
+		statementEachTiddler.reset();
+	};
+	/*
+	We get all the rows where either the shadow field is 1 and there is no row with the same title and
+	a shadow field value of zero, or the shadow field is zero and there also exists a row with the same
+	title and a shadow field value of 1
+	*/
+	var statementEachShadowTiddler = self.db.prepare(`
+			select title, meta, text
+			from tiddlers t1
+			where t1.shadow = 1
+				and not exists (
+					select 1
+					from tiddlers t2
+					where t2.title = t1.title
+						and t2.shadow = 0
+				)
+		union
+			select title, meta, text
+			from tiddlers t3
+			where t3.shadow = 0
+				and exists (
+					select 1
+					from tiddlers t4
+					where t4.title = t3.title
+						and t4.shadow = 1
+				);
```

Comment:
This looks like a syntax error with the semicolon there.

Also, I've rarely seen use of exists and not exists in SQL. The not exists can be replaced with left outer join and a normal join for exists like this:

```sql
/* shadow tiddlers which do not have any overriding tiddler */
select t1.title as title, t1.meta as meta, t1.text as text
from tiddlers t1
left join tiddlers t2 on t1.title = t2.title and t2.shadow = 0
where t1.shadow = 1 and t2.title is null
union
/* normal tiddlers which are overriding a shadow tiddler */
select t3.title, t3.meta, t3.text
from tiddlers t3
join tiddlers t4 on t3.title = t4.title and t4.shadow = 1
where t3.shadow = 0
order by title;
```
Not sure if you had specific reason to go with the exist/not exists approach, but I thought I'd share this alternative.

Reply:
> This looks like a syntax error with the semicolon there.

Thanks, fix coming up.

> Also, I've rarely seen use of exists and not exists in SQL. The not exists can be replaced with left outer join and a normal join for exists like this:

I've never really got to grips with complex SQL and so I'm ashamed to say that ChatGPT helped with these queries. It's probably related to the wording of my prompt: "get all the rows where either the shadow field is 1 and there is no row with the same title and a shadow field value of zero, or the shadow field is zero and there also exists a row with the same title and a shadow field value of 1". I think perhaps my use of the word "exists" triggered the use of that construction.

In terms of clarity, as a non-SQL person, I find the exists/not exists fairly readable, but that probably is because I need to get more familiar with the principles of relational algebra. It would be interesting to know which has better performance.

> Also, it might be worthwhile to create views for these long queries? Just an idea.

That makes sense. I've little experience with SQL since Microsoft Access days and so would welcome help on this.

Reply:
> In terms of clarity, as a non-SQL person, I find the exists/not exists fairly readable, but that probably is because I need to get more familiar with the principles of relational algebra. It would be interesting to know which has better performance.

I've written a lot of complex SQL over the years, but only sporadically and much less in recent years. By no means am I a SQL expert. A quick google search seemed to indicate that not exists might perform better. Most search results I found were related to microsoft sql server. Probably the only way to know for sure is to test, if you ever get to that point of needing to know.

> That makes sense. I've little experience with SQL since Microsoft Access days and so would welcome help on this.

This part is simple. It is just `create view <yourviewname> as <your sql query>`